packages/server/src/models/Locker.ts

```diff
@@ -0,0 +1,106 @@
+import {ERRORS} from '../constants'
+
+export type RequestRelease = () => Promise<void> | void
+
+// The Locker interface defines methods for implementing a locking mechanism.
+// This is crucial for ensuring exclusive access to uploads and their metadata.
+// Following TUS recommendations, it's important to cancel locks from previous requests
+// to avoid holding locks for too long and to manage half-open TCP connections effectively.
+// The lock method includes a cancel callback to facilitate the cancellation of a request that previously acquired the lock.
+export interface Locker {
+  lock(id: string, cancelReq: RequestRelease): Promise<void>
```

Comment:
In tusd, we don't pass the upload ID directly to `lock`/`unlock`, but instead let the locker create a lock instance, which is then used for locking and releasing: https://github.com/tus/tusd/blob/c856f723619a0a388c619d68357b9510c8632640/pkg/handler/datastore.go#L163-L190

The idea is that the lock instance can then hold references to the lock object in external services. This might be more challenging if we only operate on the upload ID.

Reply:
It seems like such a lock instance is already used by the memory locker internally.

Reply:
it's more of an implementation detail the behaviour is the same and i find the interface easier to implement by only having `lock/unlock`

Reply:
Yes, it is easier to implement it by passing around upload IDs as strings. But that interface is also more restricting because there is no state is preserved between locking and unlocking. Any state that should be preserved must be stored in a central map by the locker itself, which is not ideal. Having a lock instance with its methods for locking and unlock resolves this problem because the state can be kept in the instance itself.

I think in the future we would run into issues with the current interface, but right now we have the possibility to define a more powerful and future-proof interface.

Reply:
In any case, the locker needs to keep a map of ID -> Lock as is also implemented here:
https://github.com/tus/tusd/blob/main/pkg/memorylocker/memorylocker.go#L29

By having the Locker simply return an instance of `Lock` which then the Lock interacts with the Locker I find it more complicated and don't yet see the usefulness that can bring in the future